### PR TITLE
Add near-live refresh for CVE feed page

### LINKED ISSUE #43968: CVE feed should update in near real time
As a follow up to https://github.com/kubernetes/sig-security/pull/106, the Kube website should be re-built more often or automatically triggered so that when a CVE issue is added to the feed, it shows up on the website quickly.

cc @kubernetes/security-response-committee @kubernetes/sig-security @PushkarJ

### PR DESCRIPTION
**Pull request by ahmedavid**
Edited cve-feed.html to add script that fetches cve json feed and updates cve table periodically.

### CODE REVIEW

**Comment by sftim** [comment]
/retitle Add near-live refresh for CVE feed page

**Review by sftim — DISMISSED** [review]
Thank you for the PR.


This does not close https://github.com/kubernetes/website/issues/43968 (there will still be more work to do, around generating the JSON feed data more frequently)

Please update the PR description to reflect that @ahmedavid

**Comment by ahmedavid** [comment]
Hi @sftim , I changed the description. First part was straight forward. But I have no clue about second part. I am happy to help but I need some pointers.

**Comment by sftim** [comment]
Thanks for updating the PR description. If / when this is reviewed and approved, I expect we'll be able to merge this change without waiting on an update that addresses the other part (generating the feed data more often).

**Comment by ahmedavid** [comment]
I have couple of concerns

1. Initially I embedded <script>...</script> directly inside cve-feed short code. Maybe better approach is to extract it to static folder where other js are. Also we can put defer attribute to run it after page is rendered. Should I bother, because this works?!
2. Currently, in preview deployment implementation doesn't work because of CORS problem. Because origin of json is kubernetes.io but we are serving preview app on netlify. It is going to work on production server. To make it work on netlify we need to add CORS headers on netlify domain as well. CORS is a server side thing I am not sure how to handle server side on this project. My best guess is we need to change something in netlify.toml
3. Administrative concerns is, do I need to pursue reviewers/approvers or they will show up eventually?!

File: layouts/shortcodes/cve-feed.html [hunk 1]
```diff
@@ -21,3 +22,35 @@
...
+          <td><a href="${cve.url}">${cve.id}</a></td>
+          <td>${cve.summary}</td>
+          <td><a href="${cve.url}">#${cve["_kubernetes_io"].issue_number}</a></td>
+        </tr>`
+      } 
+      return tbodyHTML
+    } 
+
+    setInterval(() => {
+      fetch(`${window.location.origin}/docs/reference/issues-security/official-cve-feed/index.json`)
```
Comment (sftim) [inline layouts/shortcodes/cve-feed.html:45]: Here's a problem: the HTML and the JSON data at that URL get updated in lock step. So this change by itself wouldn't cut down the latency of updates.

What I think we actually want to fetch is `.Site.Params.cveFeedBucket`. I'd do that on page load and then once every 10 minutes.

Comment (ahmedavid) [reply layouts/shortcodes/cve-feed.html]: Hi @sftim , I implemented your recommendations.

**Comment by sftim** [comment]
This kind of works; the code is not _wrong_ but, it's still not the whole solution.

Actually, there's a different thing we can do. We can skip all this change and instead trigger a whole-site deploy whenever there is a diff between the public CVE feed and the source data in GCP. And some Kubernetes infrastructure could do that fix once every 10 minutes.

**Comment by ahmedavid** [comment]
> This kind of works; the code is not _wrong_ but, it's still not the whole solution.
> 
> Actually, there's a different thing we can do. We can skip all this change and instead trigger a whole-site deploy whenever there is a diff between the public CVE feed and the source data in GCP. And some Kubernetes infrastructure could do that fix once every 10 minutes.

Can you point me to the code that generates json?

Redeploy website every 10 mins? That doesn't sound like a good idea to me.

**Comment by sftim** [comment]
> Redeploy website every 10 mins? That doesn't sound like a good idea to me.

Every 10 mins:
- check if https://k8s.io/docs/reference/issues-security/official-cve-feed/index.json is out of date (by comparing it to the data in the GCP object storage)
- if so: trigger a site rebuild

This will trigger approximately 10 rebuilds a year, which is fine. It will also add a few HTTP GETs to our served traffic (which is about 10 TB / month IIRC).

See https://github.com/kubernetes/sig-security/issues/1 for the overall issue tracker and more context.

**Comment by sftim** [comment]
misclicked!
/reopen

**Comment by ahmedavid** [comment]
> > Redeploy website every 10 mins? That doesn't sound like a good idea to me.
> 
> Every 10 mins:
> 
> * check if https://k8s.io/docs/reference/issues-security/official-cve-feed/index.json is out of date (by comparing it to the data in the GCP object storage)
> * if so: trigger a site rebuild
> 
> This will trigger approximately 10 rebuilds a year, which is fine. It will also add a few HTTP GETs to our served traffic (which is about 10 TB / month IIRC).
> 
> See [kubernetes/sig-security#1](https://github.com/kubernetes/sig-security/issues/1) for the overall issue tracker and more context.

And where this logic would run?

**Comment by sftim** [comment]
> Reading through the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-security/3203-auto-refreshing-official-cve-feed) I can't tell if there is some kind of script exists that implements [Design Details](https://github.com/kubernetes/enhancements/tree/master/keps/sig-security/3203-auto-refreshing-official-cve-feed#design-details) steps 1,2,3,4. Or they put a one time static file into the bucket and that is what we are getting. If such script or automation does not exist. First that part needs to be done. Can you confirm if that part is already done?

I'm going to step back from this conversation  (as I have around 80 other things I'm also involved in for Kubernetes, and there is an upcoming release); I hope other contributors step in.
You could also ask in #sig-docs and #sig-security on the Kubernetes Slack workspace @ahmedavid - see https://slack.k8s.io/ if you are not already a user there.

**Comment by ahmedavid** [comment]
> > Reading through the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-security/3203-auto-refreshing-official-cve-feed) I can't tell if there is some kind of script exists that implements [Design Details](https://github.com/kubernetes/enhancements/tree/master/keps/sig-security/3203-auto-refreshing-official-cve-feed#design-details) steps 1,2,3,4. Or they put a one time static file into the bucket and that is what we are getting. If such script or automation does not exist. First that part needs to be done. Can you confirm if that part is already done?
> 
> I'm going to step back from this conversation (as I have around 80 other things I'm also involved in for Kubernetes, and there is an upcoming release); I hope other contributors step in. You could also ask in #sig-docs and #sig-security on the Kubernetes Slack workspace @ahmedavid - see https://slack.k8s.io/ if you are not already a user there.

Appreciate your input so far, thanks.

File: layouts/shortcodes/cve-feed.html [hunk 2]
```diff
@@ -21,3 +22,43 @@
   {{ end }}
   </tbody>
 </table>
+
+<script>
```
Comment (dipesh-rawat) [inline layouts/shortcodes/cve-feed.html:26]: Consider incorporating this script into the header or footer of the page, loading conditionally based on the presence of the `HasShortcode` parameter. 

For guidance, refer to the tips provided at https://gohugo.io/templates/shortcode-templates/ and refer our existing layouts to see how we handle 'HasShortcode'.

File: layouts/shortcodes/cve-feed.html [hunk 3]
```diff
@@ -21,3 +22,43 @@
...
+  window.addEventListener("DOMContentLoaded", () => {
+    const table = document.querySelector("table.security-cves")
+    const cveFeedBucket = table.getAttribute("data-cve-feed-bucket")
+    const tbody = table.querySelector("tbody")
+    const cachedInnerHTML = tbody.innerHTML
+  
+    const renderCVEs = (cves) => {
+      let tbodyHTML = ""
+      for (let cve of cves) {
+        tbodyHTML += `<tr>
```
Comment (dipesh-rawat) [inline layouts/shortcodes/cve-feed.html:36]: Additionally, we should also refresh the table _caption_ when fetching new data, as it provides the timestamp indicating the last update of the CVE feed data.

File: layouts/shortcodes/cve-feed.html [hunk 4]
```diff
@@ -21,3 +22,43 @@
   {{ end }}
   </tbody>
 </table>
+
+<script>
+  window.addEventListener("DOMContentLoaded", () => {
```
Comment (dipesh-rawat) [inline layouts/shortcodes/cve-feed.html:27]: 💭 How about adding a refresh/fetch button in the corner of the table, alongside this auto-trigger logic of every 10 minutes? This button could be used explicitly to refresh the feed table data.

File: layouts/shortcodes/cve-feed.html
Comment (dipesh-rawat) [inline layouts/shortcodes/cve-feed.html:1]: In the [deploy preview (here)](https://deploy-preview-44074--kubernetes-io-main-staging.netlify.app//docs/reference/issues-security/official-cve-feed/), I've see following a CORS policy error:

```console
Access to fetch at 'https://storage.googleapis.com/k8s-cve-feed/official-cve-feed.json' from origin 'https://deploy-preview-44074--kubernetes-io-main-staging.netlify.app' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

I'm hoping that this cross-site script error will be gone in the production build at  [kubernetes.io](https://kubernetes.io/)

**Comment by ericsmalling** [comment]
I'm not qualified to weigh in on the code changes here - is there anyone researching the review comments?

**Comment by sftim** [comment]
@ahmedavid thanks for the PR

As this hasn't made progress in some time, I will close it. You are welcome to either send in a new updated PR, or to reopen this one.

/close

**Pull request closed without merging**